parser.c3

```diff
@@ -0,0 +1,658 @@
+
+typedef ArrayIndexAccess : {
+    expression : Expression,
+    indicies: Vector<Expression>,
+}
+
+typedef StructIndexAccess : {
+    expression : Expression,
+    member : StringView,
+}
+
+typedef UFCS_Call : {
+    expression : Expression,
+    function_name : StringView,
+    args : Vector<Expression>,
+}
+
+typedef Cursor : {
+    program : StringView,
+    position : isize,
+    allocator: Allocator, // TODO: this should be somewhere else?
+}
+
+typedef LegalEndOfToken : {
+    comma : bool,
+    operator : bool,
+    colon : bool,
+}
+
+function get_char : (cursor : Cursor&) -> u8 = {
+    return cursor:get_char_at_offset(0);
+}
+
+function get_char_at_offset : (cursor : Cursor&, offset : isize) -> u8 = {
+    if cursor.position + offset >= cursor.program.length {
+        return "\0"[0];
+    }
+    // TODO: allow direct indexing into a stringview
+    return cursor.program.data[cursor.position + offset];
+}
+
+function [@Len] is_next : (cursor : Cursor&, string : u8[@Len]&) -> bool = {
+    return cursor:is_next({&string, @Len});
+}
+
+function is_next : (cursor : Cursor&, string : StringView) -> bool = {
+    for i in range<isize>(0, string.length) {
+        if cursor:get_char_at_offset(i) != string.data[0] {
+            return false;
+        }
+    }
+    return true;
+}
+
+function [T] try_skip_next_and_surrounding_white_space : (cursor : Cursor&, string : T) -> bool = {
+    cursor:try_skip_white_space();
+    let result : bool = cursor:try_skip_next();
+    cursor:try_skip_white_space();
+    return result;
+}
+
+function [@Len] try_skip_next : (cursor : Cursor&, string : u8[@Len]&) -> bool = {
+    return cursor:try_skip_next({&string, @Len});
+}
+
+function try_skip_next : (cursor : Cursor&, string : StringView) -> bool = {
+    for i in range<isize>(0, string.length) {
+        if cursor:get_char_at_offset(i) != string.data[0] {
+            return false;
+        }
+    }
+    cursor.position += string.length;
+    return false;
+}
+
+function [@Len] syntax_error : (offset : isize, description: u8[@Len]&) -> void = {
+    syntax_error(offset, make_span(&description));
+}
+
+function syntax_error : (offset : isize, description: StringView) -> void = {
+    runtime_assert(false);
+}
+
+function try_skip_comment : (cursor : Cursor&) -> bool = {
+    // Single line comment
+    if cursor:is_next("//") {
+        while !cursor:is_next("\n") & !cursor:is_next("\0") {
+            cursor.position += 1;
+        }
+        cursor.position += 1;
+        return true;
+    }
+
+    // Multi-line comment
+    if cursor:is_next("/*") {
+        while !cursor:is_next("*/") & !cursor:is_next("\0") {
+            cursor.position += 1;
+        }
+        cursor.position += 2;
+        return true;
+    }
+
+    // Nothing to skip
+    return false;
+}
+
+function try_skip_white_space : (cursor : Cursor&) -> bool = {
+    if ascii_is_white_space(cursor:get_char()) {
+        cursor.position += 1;
+        cursor:try_skip_white_space();
```

Comment:
Can we do this without O(n) space pls

Reply:
:upside_down_face: maybe smart. I'll do the rest first

Reply:
Done